aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,356 @@
+    return old;
+  }
+
+  bool operator==(const ITensorListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ITENSORLIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const ITensorListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+
+/*
+ * [Note: ITensorList]
+ * Wrapper around boxed and unboxed API containers.
+ *
+ * Tagged union of both API containers:
+ *     - `TensorList`, a.k.a. `ArrayRef<Tensor>` (the unboxed API container)
+ *     - `List<Tensor>` (the boxed API container)
+ *
+ * This container wraps around these two, without incurring in extra overhead
+ * for converting from one to another.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ */
+class ITensorList {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...) friend class TORCH_ITENSORLIST_IMPL(TAG);
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_type = TORCH_ITENSORLIST_IMPL(Unboxed)::list_type;
+  using boxed_type = TORCH_ITENSORLIST_IMPL(Boxed)::list_type;
+
+  union Payload {
+    const boxed_type* boxed;
+    unboxed_type unboxed;
+    Payload() : boxed(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  using iterator = ITensorListIterator;
+  using const_iterator = ITensorListIterator;
+  using value_type = typename iterator::value_type;
+
+  ITensorList() : tag_(ITensorListTag::None) {}
+
+  ITensorList(const std::initializer_list<at::Tensor>& list)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = at::ArrayRef<at::Tensor>(list);
+  }
+
+  ITensorList(const boxed_type& boxed) : tag_(ITensorListTag::Boxed) {
+    payload_.boxed = &boxed;
+  }
+
+  ITensorList(const unboxed_type& unboxed) : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed;
+  }
+
+  template <
+      typename... UnboxedConstructorArgs,
+      typename = std::enable_if_t<
+          std::is_constructible<unboxed_type, UnboxedConstructorArgs...>::value>>
+  ITensorList(UnboxedConstructorArgs&&... args)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed_type(std::forward<UnboxedConstructorArgs>(args)...);
+  }
+
+  size_t size() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.size(); });
+  }
+
+  bool empty() const {
+    return size() == 0;
+  }
+
+  iterator begin() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.begin(); });
+  }
+
+  iterator end() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.end(); });
+  }
+
+  detail::ITensorListConstRef operator[](size_t i) const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_[i]; });
+  }
+
+#define DEFINE_CHECK(TAG, ...)          \
+  bool is##TAG() const {                \
+    return tag_ == ITensorListTag::TAG; \
+  }
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_CHECK);
+#undef DEFINE_CHECK
```

Comment:
Hmm, I guess this is used for testing? It's hard to imagine a situation where the user will want it.

Reply:
Not only for that. But, it's also useful in a situation where you need exactly one of the container types. I've just bumped into this while creating an intermediate PR for modifying the generated code for structured kernels only.